specification/cognitiveservices/OpenAI.Inference/models/completions.create.tsp

```diff
@@ -1,177 +1,226 @@
+    resource URI that's connected to.
+    """)
+    @projectedName("json", "model")
+    @projectedName("csharp", "InternalNonAzureModelName")
+    "model"?: string;
 };
 
-@doc("Expected response schema to completion request")
+@doc("""
+Representation of the response data from a completions request.
+Completions support a wide variety of tasks and generate text that continues from or "completes"
+provided prompt data.
+""")
 model Completions {
-    @doc("Request ID for troubleshooting purposes")
-    @header "apim-request-id": string;
-    @doc("Id for completion response")
-    id?: string;
-    @doc("Object for completion response")
-    object: "text_completion";
-    @doc("Created time for completion response")
-    created?: int32;
-    @doc("Model used for completion response")
-    "model"?: string;
-    @doc("Array of choices returned containing text completions to prompts sent")
+    @doc("A unique identifier associated with this completions response.")
+    @projectedName("json", "id")
+    id: string;
+
+    @doc("""
+    The first timestamp associated with generation activity for this completions response,
+    represented as seconds since the beginning of the Unix epoch of 00:00 on 1 Jan 1970.
+    """)
+    @projectedName("json", "created")
+    @projectedName("csharp", "InternalCreatedSecondsAfterUnixEpoch")
+    created: int32;
+
+    @doc("""
+    The collection of completions choices associated with this completions response.
+    Generally, `n` choices are generated per provided prompt with a default value of 1.
+    Token limits and other settings may limit the number of choices generated.
+    """)
+    @projectedName("json", "choices")
     choices?: Choice[];
-    @doc("Usage counts for tokens input using the completions API")
+
+    @doc("""
+    Usage information for tokens processed and generated as part of this completions operation.
+    """)
+    @projectedName("json", "usage")
     usage: CompletionsUsage;
 }
 
-@doc("Choice model within completion response")
+@doc("""
+The representation of a single prompt completion as part of an overall completions request.
+Generally, `n` choices are generated per provided prompt with a default value of 1.
+Token limits and other settings may limit the number of choices generated.
+""")
 model Choice {
-    @doc("Generated text for given completion prompt")
-    text?: string;
-    @doc("Index")
-    index?: int32;
-    @doc("Log Prob Model")
-    logprobs?: CompletionsLogProbs;
+    @doc("The generated text for a given completions prompt.")
+    @projectedName("json", "text")
+    text: string;
+
+    @doc("The ordered index associated with this completions choice.")
+    @projectedName("json", "index")
+    index: int32;
+
+    @doc("The log probabilities model for tokens associated with this completions choice.")
+    @projectedName("json", "logprobs")
+    @projectedName("csharp", "LogProbabilityModel")
+    logprobs?: CompletionsLogProbabilityModel | null;
```

Comment:
Just double-checking: This is both, optional (the property can be *not* present in the JSON, aka undefined) and nullable (when the property is present, it can explicitly be `null`)?

Reply:
I'm glad you're asking! I didn't catch it before, but this may be a lingering emission issue.

To your question: no, I don't believe it's both--the field is seeming *always* present in /completions choices, but can be either `null` or the appropriate object.

But if I remove the parameter optionality (`logprobs?` to `logprobs`), the codegen constructor:

- Deserialization still skips over the property (and leaves the object for initialization null)
- The constructor gains a null check that trips up on that (null) item

Is that a problem with codegen (both of those are non-custom) or a usage problem with how I'm defining it?

```
logprobs: CompletionsLogProbabilityModel | null
```

Reply:
This one was discussed on the C# PR too. It is a problem with codegen and, based on that discussion, it sounds like updating to the latest fixes it, so the change can be made here.

Reply:
That's exactly it! After that offline conversation, I've updated the definition here (in the typespec files) to represent reality, and moved the minimal constructor code needed to work around the emitter issue into Custom.